share/eds/blockstore.go

```diff
@@ -0,0 +1,210 @@
+)
+
+type accessorWithBlockstore struct {
+	sa *dagstore.ShardAccessor
+	bs dagstore.ReadBlockstore
+}
+
+// Blockstore implements the blockstore.Blockstore interface on an EDSStore.
+// The lru cache approach is heavily inspired by the open PR filecoin-project/dagstore/116.
+// The main differences to the implementation here are that we do not support multiple shards per
+// key, call GetSize directly on the underlying RO blockstore, and do not throw errors on
+// Put/PutMany. Also, we do not abstract away the blockstore operations.
+//
+// The intuition here is that each CAR file is its own blockstore, so we need this top level
+// implementation to allow for the blockstore operations to be routed to the underlying stores.
+type Blockstore struct {
+	store *Store
+
+	// bsStripedLocks prevents simultaneous RW access to the blockstore cache for a shard. Instead
+	// of using only one lock or one lock per key, we stripe the shard keys across 256 locks. 256 is
+	// chosen because it 0-255 is the range of values we get looking at the last byte of the key.
+	bsStripedLocks [256]sync.Mutex
+	// caches the blockstore for a given shard for shard read affinity i.e.
+	// further reads will likely be from the same shard. Maps (shard key -> blockstore).
+	blockstoreCache *lru.Cache
+}
+
+func NewEDSBlockstore(s *Store) (*Blockstore, error) {
+	// instantiate the blockstore cache
+	bslru, err := lru.NewWithEvict(maxCacheSize, func(_ interface{}, val interface{}) {
+		// ensure we close the blockstore for a shard when it's evicted from the cache so dagstore can gc
+		// it.
+		abs, ok := val.(*accessorWithBlockstore)
+		if ok {
+			abs.sa.Close()
+		}
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate blockstore cache: %w", err)
+	}
+	return &Blockstore{
+		store:           s,
+		blockstoreCache: bslru,
+	}, nil
+}
+
+func (bs *Blockstore) Has(ctx context.Context, cid cid.Cid) (bool, error) {
+	keys, err := bs.store.dgstr.ShardsContainingMultihash(ctx, cid.Hash())
+	if err != nil {
+		return false, fmt.Errorf("failed to find shards containing multihash: %w", err)
+	}
+	return len(keys) > 0, nil
+}
+
+func (bs *Blockstore) Get(ctx context.Context, cid cid.Cid) (blocks.Block, error) {
+	blockstr, err := bs.getReadOnlyBlockstore(ctx, cid)
+	if err != nil {
+		log.Debugf("failed to get blockstore for cid %s: %s", cid, err)
+		// nmt's GetNode expects an ipld.ErrNotFound when a cid is not found.
+		return nil, ipld.ErrNotFound{Cid: cid}
+	}
+	return blockstr.Get(ctx, cid)
+}
+
+func (bs *Blockstore) GetSize(ctx context.Context, cid cid.Cid) (int, error) {
+	blockstr, err := bs.getReadOnlyBlockstore(ctx, cid)
+	if err != nil {
+		return 0, err
+	}
+	return blockstr.GetSize(ctx, cid)
+}
+
+// DeleteBlock is a noop on the EDS Blockstore that returns an ErrUnsupportedOperation when called.
+func (bs *Blockstore) DeleteBlock(context.Context, cid.Cid) error {
+	return ErrUnsupportedOperation
```

Comment:
Unfortunately no, because some of the operations (Put and PutMany) are called and aren't allowed to error. I need to find the part of code that causes this again, IIRC it's in a part of bitswap where they also comment that it's a weird behavior.

Also I am against panicking where we don't have to - the cast down below is probably okay to panic on